File: backend/src/main/java/ca/metricalsky/winston/entity/fetch/FetchOperationEntity.java
package ca.metricalsky.winston.entity.fetch;

import jakarta.persistence.Basic;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.OffsetDateTime;

@Entity
@Table(name = "fetch_operations")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class FetchOperationEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Column(name = "fetch_request_id")
    private Long fetchRequestId;

    @Basic(optional = false)
    @Enumerated(EnumType.STRING)
    @Column(name = "operation_type")
    private Type operationType;

    @Basic(optional = false)
    @Column(name = "object_id")
    private String objectId;

    @Column(name = "mode")
    private String mode;

    @Column(name = "published_after")
    private OffsetDateTime publishedAfter;

    @Column(name = "published_before")
    private OffsetDateTime publishedBefore;

    @Builder.Default
    @Basic(optional = false)
    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    private Status status = Status.READY;

    @Column(name = "error")
    private String error;

    @CreationTimestamp
    @Column(name = "created_at")
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "last_updated_at")
    private OffsetDateTime lastUpdatedAt;

    public enum Type {
        CHANNELS,
        VIDEOS,
        COMMENTS,
        REPLIES,
    }

    public enum Status {
        READY,
        FETCHING,
        SUCCESSFUL,
        FAILED,
    }
}

File: database/src/main/kotlin/ca/metricalsky/winston/database/entity/fetch/FetchOperationEntity.kt
package ca.metricalsky.winston.database.entity.fetch

import jakarta.persistence.Basic
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.OffsetDateTime

@Entity
@Table(name = "fetch_operations")
class FetchOperationEntity(

    @Column(name = "fetch_request_id")
    var fetchRequestId: Long? = null,

    @Basic(optional = false)
    @Enumerated(EnumType.STRING)
    @Column(name = "operation_type")
    var operationType: Type,

    @Basic(optional = false)
    @Column(name = "object_id")
    var objectId: String,

    @Column(name = "mode")
    var mode: String? = null,

    @Column(name = "published_after")
    var publishedAfter: OffsetDateTime? = null,

    @Column(name = "published_before")
    var publishedBefore: OffsetDateTime? = null,
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Long? = null

    @Basic(optional = false)
    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    var status = Status.READY

    @Column(name = "error")
    var error: String? = null

    @CreationTimestamp
    @Column(name = "created_at")
    var createdAt: OffsetDateTime? = null

    @UpdateTimestamp
    @Column(name = "last_updated_at")
    var lastUpdatedAt: OffsetDateTime? = null

    enum class Type {
        CHANNELS,
        VIDEOS,
        COMMENTS,
        REPLIES,
    }

    enum class Status {
        READY,
        FETCHING,
        SUCCESSFUL,
        WARNING,
        FAILED,
    }
}
